Userland/Services/FileSystemAccessServer/DownloadWidget.cpp

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (c) 2020, Andreas Kling <kling@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "DownloadWidget.h"
+#include <AK/NumberFormat.h>
+#include <AK/StringBuilder.h>
+#include <LibCore/File.h>
+#include <LibCore/FileStream.h>
+#include <LibCore/StandardPaths.h>
+#include <LibDesktop/Launcher.h>
+#include <LibGUI/BoxLayout.h>
+#include <LibGUI/Button.h>
+#include <LibGUI/ImageWidget.h>
+#include <LibGUI/Label.h>
+#include <LibGUI/MessageBox.h>
+#include <LibGUI/Progressbar.h>
+#include <LibGUI/Window.h>
+#include <LibProtocol/RequestClient.h>
+#include <LibWeb/Loader/ResourceLoader.h>
+#include <math.h>
+
+namespace FileSystemAccessServer {
+
+DownloadWidget::DownloadWidget(const URL& url, String const& destination_path)
+    : m_url(url)
+    , m_destination_path(destination_path)
+{
+    m_elapsed_timer.start();
+    m_download = Web::ResourceLoader::the().protocol_client().start_request("GET", url.to_string());
+    VERIFY(m_download);
+    m_download->on_progress = [this](Optional<u32> total_size, u32 downloaded_size) {
+        did_progress(total_size.value(), downloaded_size);
```

Comment:
You don't check `total_size.has_value()` here. If you are sure it always has a value, you could insert a `VERIFY(total_size.has_value())` to make it clear to others that this wasn't just overlooked.

Reply:
I'm very unsure in fact, I'm going surround `did_progress` with a check for `total_size.has_value()`